<!-- HTML header for doxygen 1.9.5-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" class="light-mode">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.13.2"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Minimac4: dosage_writer Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  extensions: ["tex2jax.js"],
  jax: ["input/TeX","output/HTML-CSS"],
});
</script>
<script type="text/javascript" async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">Minimac4
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.13.2 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){ initResizable(false); });
/* @license-end */
</script>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classdosage__writer-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">dosage_writer Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Handles writing genotype and dosage data to VCF/BCF/SAV files.  
 <a href="#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="dosage__writer_8hpp_source.html">dosage_writer.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:aae269edd0800b9cb7351e2c1b9cbb0ad" id="r_aae269edd0800b9cb7351e2c1b9cbb0ad"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#aae269edd0800b9cb7351e2c1b9cbb0ad">dosage_writer</a> (const std::string &amp;file_path, const std::string &amp;emp_file_path, const std::string &amp;sites_file_path, savvy::file::format file_format, std::uint8_t out_compression, const std::vector&lt; std::string &gt; &amp;sample_ids, const std::vector&lt; std::string &gt; &amp;fmt_fields, const std::string &amp;chromosome, float min_r2, bool is_temp)</td></tr>
<tr class="memdesc:aae269edd0800b9cb7351e2c1b9cbb0ad"><td class="mdescLeft">&#160;</td><td class="mdescRight">Construct a new <a class="el" href="classdosage__writer.html" title="Handles writing genotype and dosage data to VCF/BCF/SAV files.">dosage_writer</a>.  <br /></td></tr>
<tr class="separator:aae269edd0800b9cb7351e2c1b9cbb0ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad377d2146427b208430b2fff795b1643" id="r_ad377d2146427b208430b2fff795b1643"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#ad377d2146427b208430b2fff795b1643">merge_temp_files</a> (std::list&lt; savvy::reader &gt; &amp;temp_files, std::list&lt; savvy::reader &gt; &amp;temp_emp_files)</td></tr>
<tr class="memdesc:ad377d2146427b208430b2fff795b1643"><td class="mdescLeft">&#160;</td><td class="mdescRight">Merge temporary HDS and empirical dosage files into the final output.  <br /></td></tr>
<tr class="separator:ad377d2146427b208430b2fff795b1643"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaabce6f12a2883f57c6579a3170fdee7" id="r_aaabce6f12a2883f57c6579a3170fdee7"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#aaabce6f12a2883f57c6579a3170fdee7">merge_temp_files</a> (std::list&lt; std::string &gt; &amp;temp_file_paths, std::list&lt; std::string &gt; &amp;temp_emp_file_paths)</td></tr>
<tr class="separator:aaabce6f12a2883f57c6579a3170fdee7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af4eafee4fde6dbc3fdd84f2142a04521" id="r_af4eafee4fde6dbc3fdd84f2142a04521"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#af4eafee4fde6dbc3fdd84f2142a04521">write_dosages</a> (const <a class="el" href="classfull__dosages__results.html">full_dosages_results</a> &amp;hmm_results, const std::vector&lt; <a class="el" href="structtarget__variant.html">target_variant</a> &gt; &amp;tar_variants, const std::vector&lt; <a class="el" href="structtarget__variant.html">target_variant</a> &gt; &amp;tar_only_variants, std::pair&lt; std::size_t, std::size_t &gt; observed_range, const <a class="el" href="classreduced__haplotypes.html">reduced_haplotypes</a> &amp;full_reference_data, const savvy::region &amp;impute_region)</td></tr>
<tr class="memdesc:af4eafee4fde6dbc3fdd84f2142a04521"><td class="mdescLeft">&#160;</td><td class="mdescRight">Write imputed dosages and observed genotypes to output files.  <br /></td></tr>
<tr class="separator:af4eafee4fde6dbc3fdd84f2142a04521"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5c5837d918ab455d5cc56b33f13142f2" id="r_a5c5837d918ab455d5cc56b33f13142f2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#a5c5837d918ab455d5cc56b33f13142f2">print_mean_er2</a> (std::ostream &amp;os) const</td></tr>
<tr class="memdesc:a5c5837d918ab455d5cc56b33f13142f2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the mean empirical R² (ER2) values to an output stream.  <br /></td></tr>
<tr class="separator:a5c5837d918ab455d5cc56b33f13142f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Handles writing genotype and dosage data to VCF/BCF/SAV files. </p>
<p>This class manages writing variant information including dosages (HDS, LDS, DS), genotype calls (GT), quality metrics (R2, ER2, call score), and optional empirical dosage data. It supports temporary file merging, FORMAT/INFO field generation, and optional output of site information.</p>
<p>The class also tracks imputation accuracy statistics for ER2 calculation. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="aae269edd0800b9cb7351e2c1b9cbb0ad" name="aae269edd0800b9cb7351e2c1b9cbb0ad"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aae269edd0800b9cb7351e2c1b9cbb0ad">&#9670;&#160;</a></span>dosage_writer()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">dosage_writer::dosage_writer </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;</td>          <td class="paramname"><span class="paramname"><em>file_path</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;</td>          <td class="paramname"><span class="paramname"><em>emp_file_path</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;</td>          <td class="paramname"><span class="paramname"><em>sites_file_path</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">savvy::file::format</td>          <td class="paramname"><span class="paramname"><em>file_format</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::uint8_t</td>          <td class="paramname"><span class="paramname"><em>out_compression</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; std::string &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>sample_ids</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; std::string &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>fmt_fields</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;</td>          <td class="paramname"><span class="paramname"><em>chromosome</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float</td>          <td class="paramname"><span class="paramname"><em>min_r2</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool</td>          <td class="paramname"><span class="paramname"><em>is_temp</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Construct a new <a class="el" href="classdosage__writer.html" title="Handles writing genotype and dosage data to VCF/BCF/SAV files.">dosage_writer</a>. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">file_path</td><td>Path to main output file (VCF/BCF/SAV) </td></tr>
    <tr><td class="paramname">emp_file_path</td><td>Optional path for empirical dosage output </td></tr>
    <tr><td class="paramname">sites_file_path</td><td>Optional path for site info output </td></tr>
    <tr><td class="paramname">file_format</td><td>Output file format (VCF/BCF/SAV) </td></tr>
    <tr><td class="paramname">out_compression</td><td>Compression level for output file </td></tr>
    <tr><td class="paramname">sample_ids</td><td>List of sample IDs for the output </td></tr>
    <tr><td class="paramname">fmt_fields</td><td>List of FORMAT fields to include (GT, DS, GP, SD, HDS) </td></tr>
    <tr><td class="paramname">chromosome</td><td>Chromosome identifier for VCF header </td></tr>
    <tr><td class="paramname">min_r2</td><td>Minimum R2 threshold for output variants </td></tr>
    <tr><td class="paramname">is_temp</td><td>Whether the output is temporary (affects INFO/FORMAT fields) </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="ad377d2146427b208430b2fff795b1643" name="ad377d2146427b208430b2fff795b1643"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad377d2146427b208430b2fff795b1643">&#9670;&#160;</a></span>merge_temp_files() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool dosage_writer::merge_temp_files </td>
          <td>(</td>
          <td class="paramtype">std::list&lt; savvy::reader &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>temp_files</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::list&lt; savvy::reader &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>temp_emp_files</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Merge temporary HDS and empirical dosage files into the final output. </p>
<p>This function reads records from multiple temporary files generated by <code><a class="el" href="classdosage__writer.html" title="Handles writing genotype and dosage data to VCF/BCF/SAV files.">dosage_writer</a></code>, merges the INFO and FORMAT fields (HDS, LDS, GT), calculates allele frequencies, call scores, R2, and ER2 metrics, and writes the combined data to the final output files.</p>
<p>The function supports optional empirical output (<code>emp_out_file_</code>) and site information output (<code>sites_out_file_</code>). It also updates internal accuracy statistics for ER2 calculation.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">temp_files</td><td>A list of readers for temporary dosage files (HDS, GT). </td></tr>
    <tr><td class="paramname">temp_emp_files</td><td>A list of readers for temporary empirical dosage files (LDS, GT). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>true if merging succeeded and all files are consistent. </dd>
<dd>
false if any I/O error occurs or record/ploidy mismatches are detected.</dd></dl>
<dl class="section note"><dt>Note</dt><dd>The function assumes all temporary files have the same number of records and consistent ploidy per sample. Any mismatch will cause an error and abort.</dd>
<dd>
The merged HDS vector is used to generate the final FORMAT fields, including DS, HDS, GP, and SD. INFO fields are recalculated across all temporary files and may replace the intermediate S_X, S_XX, S_CS, and AN fields.</dd>
<dd>
For ER2 calculation, missing LOO_S_YY values are currently approximated by LOO_S_Y. This may need adjustment if missing values are available in temp files.</dd></dl>
<dl class="section warning"><dt>Warning</dt><dd>Only haploid and diploid samples are fully supported for SD/GP calculations.</dd>
<dd>
Any inconsistencies in record counts across temp files or mismatch between HDS, LDS, and GT vectors will abort the merge. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classdosage__writer_ad377d2146427b208430b2fff795b1643_icgraph.svg" width="471" height="56"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="aaabce6f12a2883f57c6579a3170fdee7" name="aaabce6f12a2883f57c6579a3170fdee7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaabce6f12a2883f57c6579a3170fdee7">&#9670;&#160;</a></span>merge_temp_files() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool dosage_writer::merge_temp_files </td>
          <td>(</td>
          <td class="paramtype">std::list&lt; std::string &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>temp_file_paths</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::list&lt; std::string &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>temp_emp_file_paths</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a5c5837d918ab455d5cc56b33f13142f2" name="a5c5837d918ab455d5cc56b33f13142f2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5c5837d918ab455d5cc56b33f13142f2">&#9670;&#160;</a></span>print_mean_er2()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void dosage_writer::print_mean_er2 </td>
          <td>(</td>
          <td class="paramtype">std::ostream &amp;</td>          <td class="paramname"><span class="paramname"><em>os</em></span></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the mean empirical R² (ER2) values to an output stream. </p>
<p>This function computes and writes the mean ER2 values across all accuracy statistic bins that have been tracked during imputation. For each bin, the mean ER2 is calculated as:</p>
<p class="formulaDsp">
\[  \text{mean ER2} = \frac{\text{er2_sum}}{\text{n_var}}
\]
</p>
<p>where <code>er2_sum</code> is the sum of empirical R² values for the bin, and <code>n_var</code> is the number of variants contributing to the bin. If a bin contains no variants (<code>n_var == 0</code>), a dot (<code>"."</code>) is printed instead.</p>
<p>The output is printed as a single line, prefixed by <code>"Mean ER2:"</code>, followed by space-separated values (or dots) for each bin.</p>
<p>Example output: </p><div class="fragment"><div class="line">Mean ER2: 0.85 0.72 . 0.90</div>
</div><!-- fragment --><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">os</td><td>The output stream to which the mean ER2 values should be written.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section note"><dt>Note</dt><dd><ul>
<li>This function does not modify internal state.</li>
<li>Requires that <code>accuracy_stats_</code> has been populated during the imputation process. </li>
</ul>
</dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classdosage__writer_a5c5837d918ab455d5cc56b33f13142f2_icgraph.svg" width="264" height="56"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="af4eafee4fde6dbc3fdd84f2142a04521" name="af4eafee4fde6dbc3fdd84f2142a04521"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af4eafee4fde6dbc3fdd84f2142a04521">&#9670;&#160;</a></span>write_dosages()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool dosage_writer::write_dosages </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classfull__dosages__results.html">full_dosages_results</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>hmm_results</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; <a class="el" href="structtarget__variant.html">target_variant</a> &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>tar_variants</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; <a class="el" href="structtarget__variant.html">target_variant</a> &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>tar_only_variants</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::pair&lt; std::size_t, std::size_t &gt;</td>          <td class="paramname"><span class="paramname"><em>observed_range</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classreduced__haplotypes.html">reduced_haplotypes</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>full_reference_data</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const savvy::region &amp;</td>          <td class="paramname"><span class="paramname"><em>impute_region</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Write imputed dosages and observed genotypes to output files. </p>
<p>This function iterates over the reference variants and target variants within the imputation region, writing imputed dosages and observed genotype data into VCF/BCF output streams. It handles both typed (observed) target variants and imputed-only variants, attaches INFO/FORMAT annotations, and optionally writes additional site and empirical dosage outputs.</p>
<p>Core responsibilities include:</p><ul>
<li>Aligning reference haplotype variants with target variants.</li>
<li>Writing imputed dosages for all reference variants in the region.</li>
<li>Handling target-only variants (typed but not in reference).</li>
<li>Writing empirical leave-one-out (LOO) dosages when available.</li>
<li>Applying QC filters (e.g., r² thresholds) before writing.</li>
<li>Outputting to multiple streams: primary, site-only, and empirical.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">hmm_results</td><td>Container with HMM imputation results, including dosages and leave-one-out (LOO) dosages for each reference variant.</td></tr>
    <tr><td class="paramname">tar_variants</td><td>List of target variants that overlap with the reference panel (typed genotypes available for comparison).</td></tr>
    <tr><td class="paramname">tar_only_variants</td><td>List of target variants not present in the reference panel (typed-only sites that need to be preserved).</td></tr>
    <tr><td class="paramname">observed_range</td><td>Pair of indices defining the [first, last) sample range for which observed genotypes should be considered (supports batch writing).</td></tr>
    <tr><td class="paramname">full_reference_data</td><td>Complete reference haplotype data (reduced representation), used to iterate through all reference variants in the region.</td></tr>
    <tr><td class="paramname">impute_region</td><td>Genomic region currently being imputed (chromosome, start, end).</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>true if the primary output file stream is still in a good state after writing all sites; false otherwise.</dd></dl>
<p>The algorithm proceeds in several phases:</p><ol type="1">
<li><b>Skip leading variants</b>: Both target and target-only iterators are advanced until they reach the start of the imputation region.</li>
<li><b>Iterate through reference panel</b>:<ul>
<li>For each reference variant:<ul>
<li>Write any target-only variants that occur before or at this position.</li>
<li>Build an output site with imputed dosages from <code>hmm_results</code>.</li>
<li>If the site matches a typed target variant:<ul>
<li>Add observed genotypes and leave-one-out dosages to INFO/FORMAT fields.</li>
<li>Optionally, write an empirical site record (typed + imputed).</li>
</ul>
</li>
<li>If the site does not match a target variant:<ul>
<li>Write only imputed dosages.</li>
</ul>
</li>
<li>Sites passing QC (<code>has_good_r2</code>) are written to the primary output stream and optionally to a site-only output stream.</li>
</ul>
</li>
</ul>
</li>
<li><b>Flush trailing target-only variants</b>:<ul>
<li>After finishing the reference loop, remaining target-only variants up to the end of the imputation region are written similarly.</li>
</ul>
</li>
</ol>
<p>Variants are written in VCF/BCF format, with:</p><ul>
<li><b>INFO fields</b> (imputation quality metrics, typed/imputed flags).</li>
<li><b>FORMAT fields</b> (dosages, genotypes, LOO dosages).</li>
</ul>
<dl class="section note"><dt>Note</dt><dd><ul>
<li>INFO fields written include imputation quality metrics (e.g., Rsq).</li>
<li>FORMAT fields written include dosages (DS) and optionally observed genotypes (GT).</li>
<li>Sites are only written if they pass quality filters (e.g., has_good_r2()).</li>
<li>Empirical outputs are written when both typed and imputed data are available.</li>
</ul>
</dd></dl>
<div class="dotgraph">
<iframe scrolling="no" frameborder="0" src="dot_inline_dotgraph_3.svg" width="535" height="975"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
 <div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classdosage__writer_af4eafee4fde6dbc3fdd84f2142a04521_cgraph.svg" width="398" height="246"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classdosage__writer_af4eafee4fde6dbc3fdd84f2142a04521_icgraph.svg" width="462" height="56"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="dosage__writer_8hpp_source.html">dosage_writer.hpp</a></li>
<li>src/<a class="el" href="dosage__writer_8cpp.html">dosage_writer.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.13.2
</small></address>
</div><!-- doc-content -->
</body>
</html>
