<!-- HTML header for doxygen 1.9.5-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" class="light-mode">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.13.2"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Minimac4: unique_haplotype_block Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  extensions: ["tex2jax.js"],
  jax: ["input/TeX","output/HTML-CSS"],
});
</script>
<script type="text/javascript" async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">Minimac4
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.13.2 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){ initResizable(false); });
/* @license-end */
</script>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classunique__haplotype__block-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">unique_haplotype_block Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Represents a block of unique haplotypes and their variants.  
 <a href="#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="unique__haplotype_8hpp_source.html">unique_haplotype.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:abae0532d646fcfaf21c7327012b80a51" id="r_abae0532d646fcfaf21c7327012b80a51"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#abae0532d646fcfaf21c7327012b80a51">compress_variant</a> (const <a class="el" href="structreference__site__info.html">reference_site_info</a> &amp;site_info, const std::vector&lt; std::int8_t &gt; &amp;alleles)</td></tr>
<tr class="memdesc:abae0532d646fcfaf21c7327012b80a51"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compress and map haplotype alleles for a new variant into the block.  <br /></td></tr>
<tr class="separator:abae0532d646fcfaf21c7327012b80a51"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad08c799f3eb6ffe2c7c5cad53f6381a9" id="r_ad08c799f3eb6ffe2c7c5cad53f6381a9"><td class="memItemLeft" align="right" valign="top">const std::vector&lt; <a class="el" href="structreference__variant.html">reference_variant</a> &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#ad08c799f3eb6ffe2c7c5cad53f6381a9">variants</a> () const</td></tr>
<tr class="memdesc:ad08c799f3eb6ffe2c7c5cad53f6381a9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the list of compressed reference variants in this block.  <br /></td></tr>
<tr class="separator:ad08c799f3eb6ffe2c7c5cad53f6381a9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3100e7e155907f588b0cd74eac77778b" id="r_a3100e7e155907f588b0cd74eac77778b"><td class="memItemLeft" align="right" valign="top">const std::vector&lt; std::int64_t &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#a3100e7e155907f588b0cd74eac77778b">unique_map</a> () const</td></tr>
<tr class="memdesc:a3100e7e155907f588b0cd74eac77778b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the mapping from haplotypes to unique columns.  <br /></td></tr>
<tr class="separator:a3100e7e155907f588b0cd74eac77778b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a551b270f09ebd18af3a6cc9b1a71dc8d" id="r_a551b270f09ebd18af3a6cc9b1a71dc8d"><td class="memItemLeft" align="right" valign="top">std::size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#a551b270f09ebd18af3a6cc9b1a71dc8d">expanded_haplotype_size</a> () const</td></tr>
<tr class="memdesc:a551b270f09ebd18af3a6cc9b1a71dc8d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the number of original haplotypes after expansion.  <br /></td></tr>
<tr class="separator:a551b270f09ebd18af3a6cc9b1a71dc8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac65c8fbf9494f04978421e0064b1e2ec" id="r_ac65c8fbf9494f04978421e0064b1e2ec"><td class="memItemLeft" align="right" valign="top">std::size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#ac65c8fbf9494f04978421e0064b1e2ec">unique_haplotype_size</a> () const</td></tr>
<tr class="memdesc:ac65c8fbf9494f04978421e0064b1e2ec"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the number of unique haplotypes represented in the block.  <br /></td></tr>
<tr class="separator:ac65c8fbf9494f04978421e0064b1e2ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac6accb25615e067cd96ce3abdf82e70f" id="r_ac6accb25615e067cd96ce3abdf82e70f"><td class="memItemLeft" align="right" valign="top">std::size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#ac6accb25615e067cd96ce3abdf82e70f">variant_size</a> () const</td></tr>
<tr class="memdesc:ac6accb25615e067cd96ce3abdf82e70f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the number of variants compressed in this block.  <br /></td></tr>
<tr class="separator:ac6accb25615e067cd96ce3abdf82e70f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9f93e57c8c7eb23db0e2c2f2f84534fe" id="r_a9f93e57c8c7eb23db0e2c2f2f84534fe"><td class="memItemLeft" align="right" valign="top">const std::vector&lt; std::size_t &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#a9f93e57c8c7eb23db0e2c2f2f84534fe">cardinalities</a> () const</td></tr>
<tr class="memdesc:a9f93e57c8c7eb23db0e2c2f2f84534fe"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the haplotype cardinalities across unique columns.  <br /></td></tr>
<tr class="separator:a9f93e57c8c7eb23db0e2c2f2f84534fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad3b893b578cd0a31bad9b21fa6fee652" id="r_ad3b893b578cd0a31bad9b21fa6fee652"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#ad3b893b578cd0a31bad9b21fa6fee652">clear</a> ()</td></tr>
<tr class="memdesc:ad3b893b578cd0a31bad9b21fa6fee652"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clears all data stored in the haplotype block.  <br /></td></tr>
<tr class="separator:ad3b893b578cd0a31bad9b21fa6fee652"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad2e7d5ff6d7e6e0860edb6521fc033c" id="r_aad2e7d5ff6d7e6e0860edb6521fc033c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#aad2e7d5ff6d7e6e0860edb6521fc033c">trim</a> (std::size_t min_pos, std::size_t max_pos)</td></tr>
<tr class="memdesc:aad2e7d5ff6d7e6e0860edb6521fc033c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Trims variants outside a specified genomic range.  <br /></td></tr>
<tr class="separator:aad2e7d5ff6d7e6e0860edb6521fc033c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a99fd739e1c87c723d3f4ef5ccdb4ef23" id="r_a99fd739e1c87c723d3f4ef5ccdb4ef23"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#a99fd739e1c87c723d3f4ef5ccdb4ef23">pop_variant</a> ()</td></tr>
<tr class="memdesc:a99fd739e1c87c723d3f4ef5ccdb4ef23"><td class="mdescLeft">&#160;</td><td class="mdescRight">Removes the last variant from the haplotype block.  <br /></td></tr>
<tr class="separator:a99fd739e1c87c723d3f4ef5ccdb4ef23"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6cece685c1b9ed1e8605d6da1605fca8" id="r_a6cece685c1b9ed1e8605d6da1605fca8"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#a6cece685c1b9ed1e8605d6da1605fca8">fill_cm</a> (<a class="el" href="classgenetic__map__file.html">genetic_map_file</a> &amp;map_file)</td></tr>
<tr class="memdesc:a6cece685c1b9ed1e8605d6da1605fca8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills the centimorgan (cM) values for all variants in the haplotype block.  <br /></td></tr>
<tr class="separator:a6cece685c1b9ed1e8605d6da1605fca8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad4a29467c6b8f80a328d6ea9ec47bd5d" id="r_ad4a29467c6b8f80a328d6ea9ec47bd5d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#ad4a29467c6b8f80a328d6ea9ec47bd5d">fill_cm_from_recom</a> (double &amp;start_cm)</td></tr>
<tr class="memdesc:ad4a29467c6b8f80a328d6ea9ec47bd5d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills missing centimorgan (cM) values for variants using recombination probabilities.  <br /></td></tr>
<tr class="separator:ad4a29467c6b8f80a328d6ea9ec47bd5d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ada3ab17532ea5baefd80fe72307846f2" id="r_ada3ab17532ea5baefd80fe72307846f2"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#ada3ab17532ea5baefd80fe72307846f2">deserialize</a> (std::istream &amp;is, int m3vcf_version, std::size_t n_haplotypes)</td></tr>
<tr class="memdesc:ada3ab17532ea5baefd80fe72307846f2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deserialize a unique haplotype block from an input stream (m3vcf format).  <br /></td></tr>
<tr class="separator:ada3ab17532ea5baefd80fe72307846f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae1dd4716bc959e1166e35a9c73086e36" id="r_ae1dd4716bc959e1166e35a9c73086e36"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#ae1dd4716bc959e1166e35a9c73086e36">deserialize</a> (savvy::reader &amp;input_file, savvy::variant &amp;var)</td></tr>
<tr class="memdesc:ae1dd4716bc959e1166e35a9c73086e36"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deserializes a unique haplotype block from a SAVVY input file and variant.  <br /></td></tr>
<tr class="separator:ae1dd4716bc959e1166e35a9c73086e36"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeec29660714f418ad463e25b691e192a" id="r_aeec29660714f418ad463e25b691e192a"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#aeec29660714f418ad463e25b691e192a">serialize</a> (savvy::writer &amp;output_file)</td></tr>
<tr class="memdesc:aeec29660714f418ad463e25b691e192a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Serializes the unique haplotype block to a SAVVY output file.  <br /></td></tr>
<tr class="separator:aeec29660714f418ad463e25b691e192a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad29df3bdc6f4c36d1549f519006c9ae7" id="r_ad29df3bdc6f4c36d1549f519006c9ae7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#ad29df3bdc6f4c36d1549f519006c9ae7">remove_eov</a> ()</td></tr>
<tr class="memdesc:ad29df3bdc6f4c36d1549f519006c9ae7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Removes "end-of-vector" markers from the unique haplotype map.  <br /></td></tr>
<tr class="separator:ad29df3bdc6f4c36d1549f519006c9ae7"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Represents a block of unique haplotypes and their variants. </p>
<p>This class stores haplotype information in a compressed form, mapping individual haplotypes to unique columns of alleles, tracking allele counts (cardinalities), and storing variant details. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a9f93e57c8c7eb23db0e2c2f2f84534fe" name="a9f93e57c8c7eb23db0e2c2f2f84534fe"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9f93e57c8c7eb23db0e2c2f2f84534fe">&#9670;&#160;</a></span>cardinalities()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const std::vector&lt; std::size_t &gt; &amp; unique_haplotype_block::cardinalities </td>
          <td>(</td>
          <td class="paramname"><span class="paramname"><em></em></span></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel inline">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the haplotype cardinalities across unique columns. </p>
<p>Each entry represents how many haplotypes map to the corresponding unique haplotype column.</p>
<dl class="section return"><dt>Returns</dt><dd>Const reference to the cardinalities vector. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_a9f93e57c8c7eb23db0e2c2f2f84534fe_icgraph.svg" width="408" height="126"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ad3b893b578cd0a31bad9b21fa6fee652" name="ad3b893b578cd0a31bad9b21fa6fee652"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad3b893b578cd0a31bad9b21fa6fee652">&#9670;&#160;</a></span>clear()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void unique_haplotype_block::clear </td>
          <td>(</td>
          <td class="paramname"><span class="paramname"><em></em></span></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Clears all data stored in the haplotype block. </p>
<p>This method resets the internal state of the haplotype block by removing all stored variants, haplotype mappings, and cardinality counts.</p>
<p>Specifically:</p><ul>
<li><code>variants_</code> is cleared, removing all variant records.</li>
<li><code>unique_map_</code> is cleared, removing all haplotype mapping indices.</li>
<li><code>cardinalities_</code> is cleared, resetting the frequency counts of haplotypes.</li>
</ul>
<p>After calling this method, the haplotype block is in an empty state and must be repopulated (e.g., via <code><a class="el" href="#ada3ab17532ea5baefd80fe72307846f2" title="Deserialize a unique haplotype block from an input stream (m3vcf format).">deserialize()</a></code>). </p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_ad3b893b578cd0a31bad9b21fa6fee652_icgraph.svg" width="967" height="414"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="abae0532d646fcfaf21c7327012b80a51" name="abae0532d646fcfaf21c7327012b80a51"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abae0532d646fcfaf21c7327012b80a51">&#9670;&#160;</a></span>compress_variant()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool unique_haplotype_block::compress_variant </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structreference__site__info.html">reference_site_info</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>site_info</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; std::int8_t &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>alleles</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Compress and map haplotype alleles for a new variant into the block. </p>
<p>This function updates the block of unique haplotypes by incorporating a new variant, compressing redundant haplotypes, and maintaining a mapping from input alleles to unique haplotype indices.</p>
<p>The first call initializes the block with the given variant. Subsequent calls attempt to map the alleles to existing haplotype columns or create new ones if necessary.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">site_info</td><td>Reference site metadata (chromosome, position, ID, alleles, error, recombination rate, cM). </td></tr>
    <tr><td class="paramname">alleles</td><td>Vector of observed alleles (per haplotype).</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>true if compression and mapping succeeded, false otherwise.</dd></dl>
<ul>
<li>If this is the <b>first variant</b>, a unique mapping is initialized:<ul>
<li>Each unique allele is assigned a haplotype column.</li>
<li>A <code>unique_map_</code> is built from haplotype indices to allele indices.</li>
<li><code>cardinalities_</code> tracks the number of haplotypes assigned per column.</li>
</ul>
</li>
<li>For <b>subsequent variants</b>:<ul>
<li>Alleles must match the size of <code>unique_map_</code>, otherwise the function fails.</li>
<li>Each haplotype is mapped back to its column via <code>unique_map_</code>.</li>
<li>If the allele matches the expected value, it is stored directly.</li>
<li>If it mismatches:<ul>
<li>Check if it matches a newly created column.</li>
<li>If not, a <b>new haplotype column</b> is created and propagated across all variants.</li>
</ul>
</li>
</ul>
</li>
<li>Special handling:<ul>
<li><code>savvy::typed_value::is_end_of_vector()</code> is used to mark missing/invalid alleles.</li>
<li>Ploidy mismatches are detected and reported as errors.</li>
<li><code>cardinalities_</code> is updated to reflect haplotype counts per column.</li>
</ul>
</li>
</ul>
<dl class="section note"><dt>Note</dt><dd><ul>
<li>Ensures that the total haplotype count matches across all variants.</li>
<li>Uses <code>assert()</code> checks to verify internal consistency.</li>
<li>Returns <code>false</code> if:<ul>
<li>The alleles vector is empty.</li>
<li>A sample ploidy mismatch is detected.</li>
<li>The haplotype size differs from the expected mapping.</li>
</ul>
</li>
</ul>
</dd></dl>
<p>Example use: </p><div class="fragment"><div class="line"><a class="code hl_class" href="classunique__haplotype__block.html">unique_haplotype_block</a> block;</div>
<div class="line"><a class="code hl_struct" href="structreference__site__info.html">reference_site_info</a> site{...};</div>
<div class="line">std::vector&lt;std::int8_t&gt; alleles = {0, 1, 0, 1};</div>
<div class="line"><span class="keywordtype">bool</span> ok = block.<a class="code hl_function" href="#abae0532d646fcfaf21c7327012b80a51">compress_variant</a>(site, alleles);</div>
<div class="ttc" id="aclassunique__haplotype__block_html"><div class="ttname"><a href="classunique__haplotype__block.html">unique_haplotype_block</a></div><div class="ttdoc">Represents a block of unique haplotypes and their variants.</div><div class="ttdef"><b>Definition</b> unique_haplotype.hpp:25</div></div>
<div class="ttc" id="aclassunique__haplotype__block_html_abae0532d646fcfaf21c7327012b80a51"><div class="ttname"><a href="#abae0532d646fcfaf21c7327012b80a51">unique_haplotype_block::compress_variant</a></div><div class="ttdeci">bool compress_variant(const reference_site_info &amp;site_info, const std::vector&lt; std::int8_t &gt; &amp;alleles)</div><div class="ttdoc">Compress and map haplotype alleles for a new variant into the block.</div><div class="ttdef"><b>Definition</b> unique_haplotype.cpp:10</div></div>
<div class="ttc" id="astructreference__site__info_html"><div class="ttname"><a href="structreference__site__info.html">reference_site_info</a></div><div class="ttdoc">Stores information about a site in the reference dataset.</div><div class="ttdef"><b>Definition</b> variant.hpp:40</div></div>
</div><!-- fragment --> <div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_abae0532d646fcfaf21c7327012b80a51_icgraph.svg" width="494" height="56"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ae1dd4716bc959e1166e35a9c73086e36" name="ae1dd4716bc959e1166e35a9c73086e36"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae1dd4716bc959e1166e35a9c73086e36">&#9670;&#160;</a></span>deserialize() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int unique_haplotype_block::deserialize </td>
          <td>(</td>
          <td class="paramtype">savvy::reader &amp;</td>          <td class="paramname"><span class="paramname"><em>input_file</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">savvy::variant &amp;</td>          <td class="paramname"><span class="paramname"><em>var</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Deserializes a unique haplotype block from a SAVVY input file and variant. </p>
<p>This function reads haplotype block data from the given <code>input_file</code> using the <code>var</code> object. It populates the <code>variants_</code>, <code>unique_map_</code>, and <code>cardinalities_</code> of the <code><a class="el" href="classunique__haplotype__block.html" title="Represents a block of unique haplotypes and their variants.">unique_haplotype_block</a></code>. The function stops reading when a "&lt;BLOCK&gt;" alt allele is encountered or the input ends.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">input_file</td><td>Reference to a <code>savvy::reader</code> object representing the input file. </td></tr>
    <tr><td class="paramname">var</td><td>Reference to a <code>savvy::variant</code> object used to read variant data. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>int Returns:<ul>
<li><code>-1</code> on I/O errors or invalid data,</li>
<li><code>0</code> if input is empty or not ready,</li>
<li><code>n+1</code> where <code>n</code> is the number of variants successfully deserialized.</li>
</ul>
</dd></dl>
<dl class="section note"><dt>Note</dt><dd>The <code>gt</code> vector of each variant is expected to match the size of <code>cardinalities_</code>. The allele count <code>ac</code> is computed using the <code>cardinalities_</code>. The first "&lt;BLOCK&gt;" allele encountered marks the end of this haplotype block. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_ae1dd4716bc959e1166e35a9c73086e36_cgraph.svg" width="418" height="56"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ada3ab17532ea5baefd80fe72307846f2" name="ada3ab17532ea5baefd80fe72307846f2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ada3ab17532ea5baefd80fe72307846f2">&#9670;&#160;</a></span>deserialize() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool unique_haplotype_block::deserialize </td>
          <td>(</td>
          <td class="paramtype">std::istream &amp;</td>          <td class="paramname"><span class="paramname"><em>is</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int</td>          <td class="paramname"><span class="paramname"><em>m3vcf_version</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::size_t</td>          <td class="paramname"><span class="paramname"><em>n_haplotypes</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Deserialize a unique haplotype block from an input stream (m3vcf format). </p>
<p>This function reads and parses a block of data from a multiple-phased VCF (m3vcf) file into the <code><a class="el" href="classunique__haplotype__block.html" title="Represents a block of unique haplotypes and their variants.">unique_haplotype_block</a></code> object. It populates internal structures such as the unique haplotype map, variant information, and cardinalities.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in,out]</td><td class="paramname">is</td><td>Input stream containing m3vcf block data. Stream state is updated depending on success or error. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">m3vcf_version</td><td>Format version of the m3vcf file (1 or 2). </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">n_haplotypes</td><td>Total number of haplotypes expected in this block.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>true</code> if deserialization was successful, <code>false</code> otherwise.</dd></dl>
<dl class="section note"><dt>Note</dt><dd>On failure, the method clears all internal data structures, marks the stream with <code>std::ios::badbit</code>, and writes an error message to <code>std::cerr</code>.</dd></dl>
<h2><a class="anchor" id="autotoc_md3"></a>
Parsing Steps</h2>
<ol type="1">
<li>Read the header line of the block, extract metadata:<ul>
<li><code>VARIANTS=&lt;N&gt;</code> → number of variants in the block.</li>
<li><code>REPS=&lt;M&gt;</code> → number of unique haplotype representatives.</li>
</ul>
</li>
<li>Parse haplotype indices (<code>unique_map_</code>) from genotype columns:<ul>
<li>For version 1: one haplotype index per column.</li>
<li>For version 2: paired haplotype indices separated by '|'.</li>
</ul>
</li>
<li>Validate that the number of parsed haplotypes matches <code>n_haplotypes</code>.</li>
<li>Build the <code>cardinalities_</code> vector (frequency of each unique haplotype).</li>
<li>Read <code>n_variants</code> subsequent lines, parsing:<ul>
<li>Chromosome, position, ID, reference/alternate alleles.</li>
<li>INFO fields (extracts <code>ERR</code> and <code>RECOM</code> values).</li>
<li>Genotypes:<ul>
<li>For version 2: encoded as run-length style offsets (<code>cols[8]</code>).</li>
<li>For version 1: raw 0/1 vector of length <code>n_reps</code>.</li>
</ul>
</li>
<li>Allele counts (<code>ac</code>) computed using <code>cardinalities_</code>.</li>
</ul>
</li>
</ol>
<h2><a class="anchor" id="autotoc_md4"></a>
Error Conditions</h2>
<ul>
<li>Invalid format (wrong number of columns, invalid integer conversions, etc.).</li>
<li>Number of haplotypes read does not match <code>n_haplotypes</code>.</li>
<li>Genotype column inconsistent with expected <code>n_reps</code>.</li>
<li>Truncated input (fewer than expected variant lines).</li>
</ul>
<p>In any of these cases, the function:</p><ul>
<li>Clears internal state (<code><a class="el" href="#ad3b893b578cd0a31bad9b21fa6fee652" title="Clears all data stored in the haplotype block.">clear()</a></code>).</li>
<li>Marks the stream as bad (<code>is.setstate(is.rdstate() | std::ios::badbit)</code>).</li>
<li>Prints an error message to <code>std::cerr</code>. </li>
</ul>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_ada3ab17532ea5baefd80fe72307846f2_cgraph.svg" width="418" height="56"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_ada3ab17532ea5baefd80fe72307846f2_icgraph.svg" width="732" height="164"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a551b270f09ebd18af3a6cc9b1a71dc8d" name="a551b270f09ebd18af3a6cc9b1a71dc8d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a551b270f09ebd18af3a6cc9b1a71dc8d">&#9670;&#160;</a></span>expanded_haplotype_size()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::size_t unique_haplotype_block::expanded_haplotype_size </td>
          <td>(</td>
          <td class="paramname"><span class="paramname"><em></em></span></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel inline">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the number of original haplotypes after expansion. </p>
<p>This corresponds to the length of the <code>unique_map_</code> vector.</p>
<dl class="section return"><dt>Returns</dt><dd>Expanded haplotype count. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_a551b270f09ebd18af3a6cc9b1a71dc8d_icgraph.svg" width="427" height="126"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a6cece685c1b9ed1e8605d6da1605fca8" name="a6cece685c1b9ed1e8605d6da1605fca8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6cece685c1b9ed1e8605d6da1605fca8">&#9670;&#160;</a></span>fill_cm()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void unique_haplotype_block::fill_cm </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classgenetic__map__file.html">genetic_map_file</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>map_file</em></span></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Fills the centimorgan (cM) values for all variants in the haplotype block. </p>
<p>This function iterates over all variants in the <code>variants_</code> vector and sets each variant's <code>cm</code> field by interpolating its genetic position using the provided <code><a class="el" href="classgenetic__map__file.html" title="A reader and interpolator for genetic map files.">genetic_map_file</a></code>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">map_file</td><td>Reference to a <code><a class="el" href="classgenetic__map__file.html" title="A reader and interpolator for genetic map files.">genetic_map_file</a></code> object used for interpolation. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_a6cece685c1b9ed1e8605d6da1605fca8_cgraph.svg" width="450" height="56"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_a6cece685c1b9ed1e8605d6da1605fca8_icgraph.svg" width="486" height="56"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ad4a29467c6b8f80a328d6ea9ec47bd5d" name="ad4a29467c6b8f80a328d6ea9ec47bd5d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad4a29467c6b8f80a328d6ea9ec47bd5d">&#9670;&#160;</a></span>fill_cm_from_recom()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void unique_haplotype_block::fill_cm_from_recom </td>
          <td>(</td>
          <td class="paramtype">double &amp;</td>          <td class="paramname"><span class="paramname"><em>start_cm</em></span></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Fills missing centimorgan (cM) values for variants using recombination probabilities. </p>
<p>This function iterates over all variants in the <code>variants_</code> vector. For any variant with a NaN <code>cm</code> value, it sets <code>cm</code> to the provided <code>start_cm</code>. If the variant has a valid <code>recom</code> value, <code>start_cm</code> is incremented using the recombination probability converted to cM via <code><a class="el" href="classrecombination.html#a442383839184ab50a5a97760136a1db6" title="Convert switch probability to genetic distance.">recombination::switch_prob_to_cm()</a></code>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start_cm</td><td>Reference to a double representing the starting centimorgan value. This value will be updated as variants are processed. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_ad4a29467c6b8f80a328d6ea9ec47bd5d_cgraph.svg" width="404" height="56"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a99fd739e1c87c723d3f4ef5ccdb4ef23" name="a99fd739e1c87c723d3f4ef5ccdb4ef23"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a99fd739e1c87c723d3f4ef5ccdb4ef23">&#9670;&#160;</a></span>pop_variant()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void unique_haplotype_block::pop_variant </td>
          <td>(</td>
          <td class="paramname"><span class="paramname"><em></em></span></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Removes the last variant from the haplotype block. </p>
<p>This function erases the most recently added variant in the <code>variants_</code> vector.</p>
<dl class="section note"><dt>Note</dt><dd>This only modifies the <code>variants_</code> container. It does not update <code>unique_map_</code> or <code>cardinalities_</code>, so use with caution if the block has been compressed or mapped. </dd></dl>

</div>
</div>
<a id="ad29df3bdc6f4c36d1549f519006c9ae7" name="ad29df3bdc6f4c36d1549f519006c9ae7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad29df3bdc6f4c36d1549f519006c9ae7">&#9670;&#160;</a></span>remove_eov()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void unique_haplotype_block::remove_eov </td>
          <td>(</td>
          <td class="paramname"><span class="paramname"><em></em></span></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Removes "end-of-vector" markers from the unique haplotype map. </p>
<p>This function scans the <code>unique_map_</code> vector and removes any elements that represent the end-of-vector (EOV) sentinel value used to indicate missing or invalid haplotypes.</p>
<p>After this operation, the <code>unique_map_</code> will contain only valid haplotype indices, and its size is reduced by the number of EOV entries removed. </p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_ad29df3bdc6f4c36d1549f519006c9ae7_icgraph.svg" width="732" height="116"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="aeec29660714f418ad463e25b691e192a" name="aeec29660714f418ad463e25b691e192a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aeec29660714f418ad463e25b691e192a">&#9670;&#160;</a></span>serialize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool unique_haplotype_block::serialize </td>
          <td>(</td>
          <td class="paramtype">savvy::writer &amp;</td>          <td class="paramname"><span class="paramname"><em>output_file</em></span></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Serializes the unique haplotype block to a SAVVY output file. </p>
<p>This function writes the current haplotype block, including its variants, unique haplotype mapping, and cardinalities, to the specified <code>output_file</code>. The first variant in the block is treated as a marker with "&lt;BLOCK&gt;" alt allele. Each variant stores allele counts (AC/AN), error probabilities, recombination rates, centimorgan positions, and haplotype genotypes (UHA).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">output_file</td><td>Reference to a <code>savvy::writer</code> object representing the output file. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>bool Returns <code>true</code> if serialization succeeded, <code>false</code> if the block is empty or on I/O errors.</dd></dl>
<dl class="section note"><dt>Note</dt><dd>The function sets the block size in <code>output_file</code> to align zstd compression blocks with m3vcf haplotype blocks. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_aeec29660714f418ad463e25b691e192a_icgraph.svg" width="494" height="107"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="aad2e7d5ff6d7e6e0860edb6521fc033c" name="aad2e7d5ff6d7e6e0860edb6521fc033c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aad2e7d5ff6d7e6e0860edb6521fc033c">&#9670;&#160;</a></span>trim()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void unique_haplotype_block::trim </td>
          <td>(</td>
          <td class="paramtype">std::size_t</td>          <td class="paramname"><span class="paramname"><em>min_pos</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::size_t</td>          <td class="paramname"><span class="paramname"><em>max_pos</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Trims variants outside a specified genomic range. </p>
<p>This function removes variants from the haplotype block that fall outside the interval <code>[min_pos, max_pos]</code>.</p>
<p>Behavior:</p><ul>
<li>If all variants are outside the range, the block is cleared entirely.</li>
<li>Otherwise, variants before <code>min_pos</code> or after <code>max_pos</code> are erased, while the remaining variants are preserved.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">min_pos</td><td>The minimum genomic position (inclusive). </td></tr>
    <tr><td class="paramname">max_pos</td><td>The maximum genomic position (inclusive).</td></tr>
  </table>
  </dd>
</dl>
<dl class="section note"><dt>Note</dt><dd>This function only modifies the <code>variants_</code> container. The mappings (<code>unique_map_</code>) and cardinalities remain unchanged, so care must be taken when trimming after compression. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_aad2e7d5ff6d7e6e0860edb6521fc033c_cgraph.svg" width="418" height="126"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_aad2e7d5ff6d7e6e0860edb6521fc033c_icgraph.svg" width="732" height="116"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ac65c8fbf9494f04978421e0064b1e2ec" name="ac65c8fbf9494f04978421e0064b1e2ec"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac65c8fbf9494f04978421e0064b1e2ec">&#9670;&#160;</a></span>unique_haplotype_size()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::size_t unique_haplotype_block::unique_haplotype_size </td>
          <td>(</td>
          <td class="paramname"><span class="paramname"><em></em></span></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel inline">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the number of unique haplotypes represented in the block. </p>
<p>This corresponds to the number of columns in the genotype matrix. If the block is empty, returns 0.</p>
<dl class="section return"><dt>Returns</dt><dd>Unique haplotype count. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_ac65c8fbf9494f04978421e0064b1e2ec_icgraph.svg" width="408" height="126"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="a3100e7e155907f588b0cd74eac77778b" name="a3100e7e155907f588b0cd74eac77778b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3100e7e155907f588b0cd74eac77778b">&#9670;&#160;</a></span>unique_map()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const std::vector&lt; std::int64_t &gt; &amp; unique_haplotype_block::unique_map </td>
          <td>(</td>
          <td class="paramname"><span class="paramname"><em></em></span></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel inline">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the mapping from haplotypes to unique columns. </p>
<p>Each entry corresponds to an original haplotype index and indicates which compressed column it maps to.</p>
<dl class="section return"><dt>Returns</dt><dd>Const reference to the unique haplotype mapping vector. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_a3100e7e155907f588b0cd74eac77778b_icgraph.svg" width="732" height="255"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ac6accb25615e067cd96ce3abdf82e70f" name="ac6accb25615e067cd96ce3abdf82e70f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac6accb25615e067cd96ce3abdf82e70f">&#9670;&#160;</a></span>variant_size()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::size_t unique_haplotype_block::variant_size </td>
          <td>(</td>
          <td class="paramname"><span class="paramname"><em></em></span></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel inline">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the number of variants compressed in this block. </p>
<dl class="section return"><dt>Returns</dt><dd>Variant count. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_ac6accb25615e067cd96ce3abdf82e70f_icgraph.svg" width="967" height="186"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="ad08c799f3eb6ffe2c7c5cad53f6381a9" name="ad08c799f3eb6ffe2c7c5cad53f6381a9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad08c799f3eb6ffe2c7c5cad53f6381a9">&#9670;&#160;</a></span>variants()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const std::vector&lt; <a class="el" href="structreference__variant.html">reference_variant</a> &gt; &amp; unique_haplotype_block::variants </td>
          <td>(</td>
          <td class="paramname"><span class="paramname"><em></em></span></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel inline">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the list of compressed reference variants in this block. </p>
<dl class="section return"><dt>Returns</dt><dd>Const reference to the vector of reference variants. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classunique__haplotype__block_ad08c799f3eb6ffe2c7c5cad53f6381a9_icgraph.svg" width="956" height="379"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="unique__haplotype_8hpp_source.html">unique_haplotype.hpp</a></li>
<li>src/<a class="el" href="unique__haplotype_8cpp.html">unique_haplotype.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.13.2
</small></address>
</div><!-- doc-content -->
</body>
</html>
