<!-- HTML header for doxygen 1.9.5-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" class="light-mode">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.13.2"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Minimac4: hidden_markov_model Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  extensions: ["tex2jax.js"],
  jax: ["input/TeX","output/HTML-CSS"],
});
</script>
<script type="text/javascript" async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">Minimac4
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.13.2 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){ initResizable(false); });
/* @license-end */
</script>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classhidden__markov__model-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">hidden_markov_model Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Implements a Hidden Markov Model for genotype imputation.  
 <a href="#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="hidden__markov__model_8hpp_source.html">hidden_markov_model.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:af5405b8d8ee2df1747c0dfe71c544ba7" id="r_af5405b8d8ee2df1747c0dfe71c544ba7"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#af5405b8d8ee2df1747c0dfe71c544ba7">hidden_markov_model</a> (float s3_prob_threshold, float s1_prob_threshold, float diff_threshold, float background_error, float decay)</td></tr>
<tr class="memdesc:af5405b8d8ee2df1747c0dfe71c544ba7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructs a Hidden Markov Model with specified parameters.  <br /></td></tr>
<tr class="separator:af5405b8d8ee2df1747c0dfe71c544ba7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa1459d3b24365049cb34ea8f40ff828d" id="r_aa1459d3b24365049cb34ea8f40ff828d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#aa1459d3b24365049cb34ea8f40ff828d">traverse_forward</a> (const std::deque&lt; <a class="el" href="classunique__haplotype__block.html">unique_haplotype_block</a> &gt; &amp;ref_haps, const std::vector&lt; <a class="el" href="structtarget__variant.html">target_variant</a> &gt; &amp;tar_variant, std::size_t hap_idx)</td></tr>
<tr class="memdesc:aa1459d3b24365049cb34ea8f40ff828d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs a forward traversal over reference haplotypes for a given target haplotype.  <br /></td></tr>
<tr class="separator:aa1459d3b24365049cb34ea8f40ff828d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa0e5498c4f7f7ed9f11d1f81e3335fc6" id="r_aa0e5498c4f7f7ed9f11d1f81e3335fc6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="#aa0e5498c4f7f7ed9f11d1f81e3335fc6">traverse_backward</a> (const std::deque&lt; <a class="el" href="classunique__haplotype__block.html">unique_haplotype_block</a> &gt; &amp;ref_haps, const std::vector&lt; <a class="el" href="structtarget__variant.html">target_variant</a> &gt; &amp;tar_variant, std::size_t hap_idx, std::size_t out_idx, const std::vector&lt; std::vector&lt; std::vector&lt; std::size_t &gt; &gt; &gt; &amp;reverse_maps, <a class="el" href="classfull__dosages__results.html">full_dosages_results</a> &amp;output, const <a class="el" href="classreduced__haplotypes.html">reduced_haplotypes</a> &amp;full_reference_data)</td></tr>
<tr class="memdesc:aa0e5498c4f7f7ed9f11d1f81e3335fc6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs a backward traversal over reference haplotypes to compute posterior probabilities.  <br /></td></tr>
<tr class="separator:aa0e5498c4f7f7ed9f11d1f81e3335fc6"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Implements a Hidden Markov Model for genotype imputation. </p>
<p>This class performs multi-stage HMM-based imputation of genotype dosages for target haplotypes using reference haplotype blocks. It maintains forward and backward probability matrices, junction proportions, and intermediate haplotype states for S1, S2, and S3 probability transformations.</p>
<ul>
<li>The model supports typed and untyped site imputation.</li>
<li>Probabilities are normalized and adjusted for recombination, background error, and leave-one-out cross-validation.</li>
<li>Implements forward traversal (<code>traverse_forward</code>) and backward traversal (<code>traverse_backward</code>) along haplotype blocks. </li>
</ul>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="af5405b8d8ee2df1747c0dfe71c544ba7" name="af5405b8d8ee2df1747c0dfe71c544ba7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af5405b8d8ee2df1747c0dfe71c544ba7">&#9670;&#160;</a></span>hidden_markov_model()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">hidden_markov_model::hidden_markov_model </td>
          <td>(</td>
          <td class="paramtype">float</td>          <td class="paramname"><span class="paramname"><em>s3_prob_threshold</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float</td>          <td class="paramname"><span class="paramname"><em>s1_prob_threshold</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float</td>          <td class="paramname"><span class="paramname"><em>diff_threshold</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float</td>          <td class="paramname"><span class="paramname"><em>background_error</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float</td>          <td class="paramname"><span class="paramname"><em>decay</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructs a Hidden Markov Model with specified parameters. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s3_prob_threshold</td><td>Threshold probability for S3 state. </td></tr>
    <tr><td class="paramname">s1_prob_threshold</td><td>Threshold probability for S1 state. </td></tr>
    <tr><td class="paramname">diff_threshold</td><td>Minimum difference required between probabilities to make a confident state call. </td></tr>
    <tr><td class="paramname">background_error</td><td>Expected background error rate. </td></tr>
    <tr><td class="paramname">decay</td><td>Decay factor controlling the influence of previous states.</td></tr>
  </table>
  </dd>
</dl>
<p>This constructor initializes the internal HMM parameters. These thresholds and the decay factor influence the model's sensitivity to differences in observed probabilities and determine how the hidden states are inferred. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="aa0e5498c4f7f7ed9f11d1f81e3335fc6" name="aa0e5498c4f7f7ed9f11d1f81e3335fc6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa0e5498c4f7f7ed9f11d1f81e3335fc6">&#9670;&#160;</a></span>traverse_backward()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void hidden_markov_model::traverse_backward </td>
          <td>(</td>
          <td class="paramtype">const std::deque&lt; <a class="el" href="classunique__haplotype__block.html">unique_haplotype_block</a> &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>ref_haps</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; <a class="el" href="structtarget__variant.html">target_variant</a> &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>tar_variant</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::size_t</td>          <td class="paramname"><span class="paramname"><em>hap_idx</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::size_t</td>          <td class="paramname"><span class="paramname"><em>out_idx</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; std::vector&lt; std::vector&lt; std::size_t &gt; &gt; &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>reverse_maps</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classfull__dosages__results.html">full_dosages_results</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>output</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classreduced__haplotypes.html">reduced_haplotypes</a> &amp;</td>          <td class="paramname"><span class="paramname"><em>full_reference_data</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Performs a backward traversal over reference haplotypes to compute posterior probabilities. </p>
<p>This function implements the backward algorithm of a Hidden Markov Model (HMM) for a given target haplotype. It computes probabilities by combining forward probabilities (from <code>traverse_forward</code>) with backward likelihoods, handling recombination events, junction proportions, and leave-one-out considerations.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ref_haps</td><td>A deque of reference haplotype blocks (<code><a class="el" href="classunique__haplotype__block.html" title="Represents a block of unique haplotypes and their variants.">unique_haplotype_block</a></code>). Each block contains multiple haplotypes and variant positions. </td></tr>
    <tr><td class="paramname">tar_variants</td><td>A vector of target variants (<code><a class="el" href="structtarget__variant.html" title="Represents a variant in the target dataset.">target_variant</a></code>) to traverse backward. </td></tr>
    <tr><td class="paramname">hap_idx</td><td>Index of the target haplotype to traverse within <code>tar_variants</code>. </td></tr>
    <tr><td class="paramname">out_idx</td><td>Output index used for storing results in <code><a class="el" href="classfull__dosages__results.html" title="Stores full and leave-one-out (LOO) dosages for imputed variants.">full_dosages_results</a></code>. </td></tr>
    <tr><td class="paramname">reverse_maps</td><td>Maps from expanded to unique haplotypes for each block. </td></tr>
    <tr><td class="paramname">output</td><td>Reference to a <code><a class="el" href="classfull__dosages__results.html" title="Stores full and leave-one-out (LOO) dosages for imputed variants.">full_dosages_results</a></code> structure where computed posterior dosages are stored. </td></tr>
    <tr><td class="paramname">full_reference_data</td><td>Reduced haplotype reference data needed for imputation.</td></tr>
  </table>
  </dd>
</dl>
<p>The function performs the following steps:</p><ol type="1">
<li>Initializes backward probability vectors (<code>backward</code>, <code>backward_norecom</code>) and junction proportions for the last block.</li>
<li>Iterates backward over all reference blocks:<ul>
<li>Computes junction probabilities using the backward pass.</li>
<li>Normalizes probabilities to ensure valid distributions.</li>
<li>Updates constants used for imputation with combined forward and backward probabilities.</li>
<li>Calls <code>impute</code> to compute posterior dosages for each variant position.</li>
<li>Conditions backward probabilities on observed target genotypes if available.</li>
</ul>
</li>
<li>Handles recombination corrections and precision jumps using <code>transpose</code>.</li>
</ol>
<dl class="section note"><dt>Note</dt><dd><ul>
<li>Assertions ensure probability values remain within valid ranges and that the global index reaches -1 at the end.</li>
<li>Observed genotypes with negative values are treated as missing and skipped during conditioning.</li>
<li>This method relies on data structures populated during <code>traverse_forward</code>. </li>
</ul>
</dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classhidden__markov__model_aa0e5498c4f7f7ed9f11d1f81e3335fc6_cgraph.svg" width="427" height="523"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<a id="aa1459d3b24365049cb34ea8f40ff828d" name="aa1459d3b24365049cb34ea8f40ff828d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa1459d3b24365049cb34ea8f40ff828d">&#9670;&#160;</a></span>traverse_forward()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void hidden_markov_model::traverse_forward </td>
          <td>(</td>
          <td class="paramtype">const std::deque&lt; <a class="el" href="classunique__haplotype__block.html">unique_haplotype_block</a> &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>ref_haps</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; <a class="el" href="structtarget__variant.html">target_variant</a> &gt; &amp;</td>          <td class="paramname"><span class="paramname"><em>tar_variant</em></span>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::size_t</td>          <td class="paramname"><span class="paramname"><em>hap_idx</em></span>&#160;)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Performs a forward traversal over reference haplotypes for a given target haplotype. </p>
<p>This function implements the forward algorithm of a Hidden Markov Model (HMM) to compute the probability of observing the target variants given the reference haplotypes. It accounts for recombination events and optional leave-one-out calculations.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ref_haps</td><td>A deque of reference haplotype blocks (<code><a class="el" href="classunique__haplotype__block.html" title="Represents a block of unique haplotypes and their variants.">unique_haplotype_block</a></code>). Each block contains multiple haplotypes and variant positions. </td></tr>
    <tr><td class="paramname">tar_variants</td><td>A vector of target variants (<code><a class="el" href="structtarget__variant.html" title="Represents a variant in the target dataset.">target_variant</a></code>) for which the forward probabilities are computed. </td></tr>
    <tr><td class="paramname">hap_idx</td><td>The index of the target haplotype to traverse within <code>tar_variants</code>.</td></tr>
  </table>
  </dd>
</dl>
<p>The function performs the following steps:</p><ol type="1">
<li>Initializes internal forward probability matrices (<code>forward_probs_</code>, <code>forward_norecom_probs_</code>) and junction probabilities (<code>junction_prob_proportions_</code>).</li>
<li>For the first reference block, likelihoods are initialized using <code>initialize_likelihoods</code>.</li>
<li>For subsequent blocks:<ul>
<li>Computes junction probabilities based on previous block probabilities and recombination proportions.</li>
<li>Normalizes junction probabilities to ensure they sum to 1.</li>
<li>Updates <code>precision_jumps_</code> by transposing probabilities to the next position.</li>
</ul>
</li>
<li>Iterates over all variants in each block:<ul>
<li>Updates probabilities conditioned on observed target genotypes using <code>condition</code>.</li>
<li>Computes forward probabilities for the next position using <code>transpose</code>.</li>
</ul>
</li>
</ol>
<dl class="section note"><dt>Note</dt><dd><ul>
<li>Assertions ensure probability values remain within valid ranges [0, 1].</li>
<li>The function supports handling missing genotypes (<code>observed &lt; 0</code>) by skipping conditioning for those positions. </li>
</ul>
</dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><iframe scrolling="no" frameborder="0" src="classhidden__markov__model_aa1459d3b24365049cb34ea8f40ff828d_cgraph.svg" width="427" height="472"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="hidden__markov__model_8hpp_source.html">hidden_markov_model.hpp</a></li>
<li>src/<a class="el" href="hidden__markov__model_8cpp.html">hidden_markov_model.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.13.2
</small></address>
</div><!-- doc-content -->
</body>
</html>
